File: list-reduction/josephus.js
/*
$ /usr/local/src/v8/d8 josephus.js
26
list reduction array
1.232
1.217
1.217
1.224
1.22
1.227
1.224
1.226
1.219
1.217
*/

function countoffSoldiers(n,kth) {
   var soldiers = []
   for (var i = 0; i < n; i++) soldiers[i] = i+1

   //var k = kth-1, survivedLastRound = n; // standard Josephus problem
   var k = 0, survivedLastRound = n

   while (survivedLastRound > 1) {
      var survived = 0

      for (var i = 0 ; i < survivedLastRound; i++)
         if (i != k) 
            soldiers[survived++] = soldiers[i]
         else
            k += kth

         k -= survivedLastRound // wrap around
         survivedLastRound = survived
      }

      return soldiers[0] 
} 


//-----------------------------------------------------------------------------

function run_iterations(iterations, times) {
    for (var t = 0 ; t < times ; t++) {
        var start = new Date()
        for(var i = 0 ; i < iterations ; i++) {
            countoffSoldiers(40,3)
        }
        var end = new Date()
        print ((end.getTime() - start.getTime()) * 1000 / ITERS)
    }
}

print(countoffSoldiers(40,3))
ITERS = 1000000
print ("list reduction array")
run_iterations(ITERS,10)

File: list-reduction/josephus.py
"""
$ python josephus.py
26
list reduction
17.4075951576
17.7173249722
17.6679799557
17.723637104
17.6835641861
17.7069940567
17.6745419502
17.7110111713
17.6747300625
17.7083261013
"""

def countoffSoldiers(n,kth):
   soldiers = list(range(1,n+1))

   #k = kth-1; survivedLastRound = n # standard Josephus problem
   k = 0; survivedLastRound = n

   while survivedLastRound > 1:
      survived = 0

      for i in range(survivedLastRound):
         if i != k: 
            soldiers[survived] = soldiers[i]
            survived += 1
         else:
            k += kth

      k -= survivedLastRound # wrap around
      survivedLastRound = survived

   return soldiers[0] 


#-----------------------------------------------------------------------------

import time
def run_iterations(iterations, times) :
    print 'list reduction' 
    for t in range(times) :
        start = time.time()
        for i in range(iterations) :
            countoffSoldiers(40,3)
        end = time.time()
        print ((end - start) * 1000000 / iterations)

print countoffSoldiers(40,3)
ITER = 1000000
run_iterations(ITER,10)

